### PR TITLE
feat!: add open() function

### PR DESCRIPTION
**Pull request by Symbitic**
This makes it possible to open a database later or after closing.
Server applications like Fresh need this because the database path might not be known at program startup.

### CODE REVIEW

File: src/database.ts [hunk 1]
```diff
@@ -136,7 +140,21 @@ export class Database {
...
+    this.unsafeConcurrency = options.unsafeConcurrency ?? false;
+  
+    if (!path) {
+      return;
+    }
+    this.open(path, options);
+  }
+
+  /** Open a Database. */
+  open(path: string | URL, options: DatabaseOpenOptions = {}): boolean {
```
Comment (DjDeveloperr) [inline src/database.ts:154]: Unsure about returning a boolean here. I think we should simply error if user tries to open an already open Database connection object, and make the return type void to be consistent with rest of the API.

Comment (Symbitic) [reply src/database.ts]: I went ahead and changed that. I also added a TypeError in case path is empty.

File: src/database.ts [hunk 2]
```diff
@@ -84,24 +84,28 @@ export function isComplete(statement: string): boolean {
...
 
   /** Whether to support BigInt columns. False by default, integers larger than 32 bit will be inaccurate. */
   int64: boolean;
 
   unsafeConcurrency: boolean;
 
-  /** Whether DB connection is open */
-  get open(): boolean {
+  /** Whether DB connection is open. */
+  get connected(): boolean {
```
Comment (DjDeveloperr) [inline src/database.ts:104]: Looks like an unexpected breaking change in API

**Comment by DjDeveloperr** [comment]
This same functionality can be achieved by creating a new `Database` object itself only when needed. Such as

```ts
// proposed solution:
const db = new Database();
// ...
db.open("...");

// why not this?
let db: Database;
// ...
db = new Database("...");
```

First one would create DB with handle as `null` pointer which could cause some undefined behavior if somehow sent to native side. I feel like the latter is better approach. What are your views on this?

**Comment by Symbitic** [comment]
That might cause problems for classes (`#db: Database!`) or classes that extend Database.
I can add a `if (!#open) throw new Error("Database not open");` error to the beginning of every method.
Should probably do that anyway since database can be closed.

**Comment by DjDeveloperr** [comment]
> That might cause problems for classes (`#db: Database!`) or classes that extend Database.

Sorry, I think I'm unable to understand the problems here. Can you give an example? I haven't seen this API pattern before.

**Comment by Symbitic** [comment]
```ts
class MyClass {
  #db: Database;

  constructor() {}
}
```
Will cause an error because Database isn't created in the constructor.
So will this:
```ts
export class MyDatabase extends Database {
  constructor() {}
}
```

**Comment by DjDeveloperr** [comment]
You can do `#db!: Database` in the former though

**Comment by Symbitic** [comment]
True, but that doesn't solve the latter.
It also wouldn't solve the issue of how to reopen a closed database.

**Comment by DjDeveloperr** [comment]
The latter is more of an alternative for the former approach.
And in case we do need to reopen a database, is there anything wrong with just creating new Database object?

**Comment by Symbitic** [comment]
Technically those are workarounds, but this is starting to feel like a hack to get around a basic feature. `open` is part of the SQLite API; why shouldn't this wrapper class have it too?

Side note: even without an `open` method, as long as there's a `close` method, every other method should add `if (!this.#open) throw new Error("Database was closed");` to the beginning to prevent accessing an invalid pointer.

**Comment by DjDeveloperr** [comment]
`sqlite3_open` function in the C-API is a constructor for `sqlite3*` object. The wrapper for it does the same right now. I haven't seen an API like this implemented in other major database drivers before so I'm bit unsure of adding it.

**Comment by Symbitic** [comment]
Do you want to close this?

**Comment by DjDeveloperr** [comment]
This API is not a common pattern among database drivers and there isn't a strong enough use-case that is not already solved without it. I'll close this for now, but if it gains traction in future, it can be reopened.

**Pull request closed without merging**